pkg/pub_package_reader/lib/src/pubspec_content_override.dart

```diff
@@ -0,0 +1,153 @@
+// Copyright (c) 2022, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:logging/logging.dart';
+import 'package:pub_semver/pub_semver.dart';
+import 'package:yaml/yaml.dart' as yaml;
+import 'package:yaml_edit/yaml_edit.dart';
+
+final _logger = Logger('pubspec_content_override');
+
+/// The version regular expression that accepted any character as separator.
+final _lenientRegExp = RegExp(r'(\d+).(\d+).(\d+)' // Version number.
+    r'(-([0-9A-Za-z-]+(\.[0-9A-Za-z-]+)*))?' // Pre-release.
+    r'(\+([0-9A-Za-z-]+(\.[0-9A-Za-z-]+)*))?' // Build.
+    );
+
+/// Packages with bad versions:
+const _packagesWithBadVersions = <String>{
+  'assets_audio_player_web',
+  'bloc_pattern',
+  'bokain_models',
+  'built_collection',
+  'built_redux',
+  'built_redux_thunk',
+  'built_value',
+  'built_value_generator',
+  'built_value_test',
+  'built_viewmodel',
+  'built_viewmodel_flutter_built_redux',
+  'built_viewmodel_flutter_redux',
+  'built_viewmodel_generator',
+  'carp_context_package',
+  'codemetrics',
+  'fluttercontactpicker',
+  'interactive_webview',
+  'meta_types',
+  'meta_types_generator',
+  'meta_types_json',
+  'meta_types_json_generator',
+  'meta_types_redux',
+  'meta_types_redux_generator',
+  'mocktail',
+  'mocktail_image_network',
+  'polymer_interop',
+};
+
+/// Override pubspec.yaml content if needed:
+///
+/// - If the archive was created before 2022, we may need to update the versions,
+///   as pub_semver accepted separator characters other than `.`.
+String overridePubspecContentIfNeeded({
+  required String content,
+  required DateTime created,
+}) {
+  // quick checks to skip overrides
+  if (created.year >= 2022) {
+    return content;
+  }
+  if (_packagesWithBadVersions.every((name) => !content.contains(name))) {
+    return content;
+  }
+
+  try {
+    final c = _tryOverrideContent(content);
+    if (c != null && c != content) {
+      _logger.info('Content override for pubspec successful.');
+    }
+    return c ?? content;
+  } on FormatException catch (e, st) {
+    _logger.info('Unable to parse pubspec.', e, st);
+  }
+  // fallback: don't override content
+  return content;
+}
+
+String? _tryOverrideContent(String content) {
```

Comment:
```suggestion
/// Fixes broken version and version constraints in [pubspecYaml] and returns a new YAML string.
///
/// Prior to [pub_semver#63] version numbers and version constraints were accepted with any
/// character as separate, instead of requiring  `.` (dot). This occurred because the regular expression
/// in `package:pub_semver` didn't escape dots as `\.`, but just contained `.` (dot, meaning any character).
///
/// Thus, package versions published prior to 2022-01-01, maybe contain _version numbers_,
/// _version constraints_ and _SDK constraints_ that doesn't use `.` (dot) as separator. As removing
/// these _package versions_ would break existing users, this function will rewrite the version numbers
/// and version constraints to have a valid format equivalent to how older versions of `pub_semver`
/// would have interpreted the version number / version constraints.
///
/// [pub_semver#63]: https://github.com/dart-lang/pub_semver/pull/63
String _fixupBrokenVersionAndConstraints(String pubspecYaml) {
```